posthog/models/historical_version.py

```diff
@@ -0,0 +1,120 @@
+import copy
+from datetime import datetime
+from typing import Dict
+
+from django.db import models
+from django.utils import timezone
+
+
+def as_deletion_state(metadata: Dict) -> Dict:
+    """
+    Deletion state is being saved without access to a DRF serializer
+    So replace datetime field with a string or it can't be serialized to JSON
+    """
+    state = copy.deepcopy(metadata)
+    if state["created_at"] and isinstance(state["created_at"], datetime):
+        state["created_at"] = state["created_at"].isoformat()
+    return state
+
+
+class HistoricalVersion(models.Model):
+    """
+    We don't store foreign key references cos the referenced model will change or be deleted.
+    The history log should hold the state at the time it is written not the time it is read
+
+    Everything in the log must have either a team id or an organization id
+    """
+
+    class Action(models.TextChoices):
+        Create = "create"
+        Update = "update"
+        Delete = "delete"
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=["organization_id", "team_id", "name", "versioned_at"], name="unique_version"
+            ),
+            models.CheckConstraint(
+                check=models.Q(team_id__isnull=False) | models.Q(organization_id__isnull=False),
+                name="must_have_team_or_organization_id",
+            ),
+        ]
+        indexes = [
+            models.Index(fields=["item_id", "team_id", "name"]),
+        ]
+
+    # JSON of the historical item
+    state = models.JSONField(null=False)
+    # e.g. FeatureFlags - in practice this will be the name of a model class
+    name = models.fields.CharField(max_length=79, null=False)
+
+    action = models.CharField(max_length=6, choices=Action.choices, blank=True, null=False)
+
+    # the id of the item being versioned
+    # this might be a numerical id, short id, or UUID, so each will be converted to string
+    # it will be used to lookup rows with exactly matching item_ids
+    # it probably only needs to be 36 characters in order to hold a GUID
+    # but 72 may be useful to avoid a migration in future
+    item_id = models.fields.CharField(max_length=72, null=False)
+
+    # to avoid an integer version field for ordering revisions
+    versioned_at: models.DateTimeField = models.DateTimeField(default=timezone.now)
+
+    # user that caused the change
+    created_by_email = models.EmailField(null=False)
+    # max length from User model
+    created_by_name = models.CharField(max_length=150, null=False)
+    created_by_id = models.PositiveIntegerField(null=False)
```

Comment:
Hm, but then if someone changes their email (let's say) the avatar may change, so that can also raise eyebrows. Also if for GDPR reasons a user asks us to remove their data, this'd be difficult to purge. The more general solution to this may be making a policy (at least for PostHog Cloud) of only _deleting_ accounts in case of GDPR requests and otherwise always just _deactivating_. Would that be viable you think?

Reply:
> Also if for GDPR reasons a user asks us to remove their data, this'd be difficult to purge.

You have ID, name, and email to join on. So it'd be an update at time of someone exercising their GDPR rights. I don't think this makes it hard... 

I have had to explain this decision three times now though. Probably better to have a solution that doesn't generate that load

Reply:
Writing out the consequences as a table was intended to help me communicate why I'd made the decision. But it made me realise that we either write some code to handle name changes or we write some code to handle deletion. (if we don't ignore deletion and allow ourselves to lose data in the history log)

I _think_ that writing some code to handle deletion is less surprising and more easily delayed. So have switched to a foreign key for the user.